src/errors.rs

```diff
@@ -1,67 +1,211 @@
-use output;
-use std::ffi::OsString;
+use std::ffi;
+use std::fmt;
+use std::io;
 
-const ERROR_PREFIX: &'static str = "CLI assertion failed";
+use failure;
 
-fn format_cmd(cmd: &[OsString]) -> String {
+fn format_cmd(cmd: &[ffi::OsString]) -> String {
     let result: Vec<String> = cmd.iter()
         .map(|s| s.to_string_lossy().into_owned())
         .collect();
     result.join(" ")
 }
 
-error_chain! {
-    links {
-        Output(output::Error, output::ErrorKind);
+pub trait ChainFail {
+    fn chain<E>(self, cause: E) -> Self
+    where
+        E: Into<failure::Error>;
+}
+
+pub trait ResultChainExt<T> {
+    fn chain<C>(self, chainable: C) -> Result<T, C>
+    where
+        C: ChainFail;
+
+    fn chain_with<F, C>(self, chainable: F) -> Result<T, C>
+    where
+        F: FnOnce() -> C,
+        C: ChainFail;
+}
+
+impl<T> ResultChainExt<T> for Result<T, failure::Error> {
+    fn chain<C>(self, chainable: C) -> Result<T, C>
+    where
+        C: ChainFail,
+    {
+        self.map_err(|e| chainable.chain(e))
     }
-    foreign_links {
-        Io(::std::io::Error);
-        Fmt(::std::fmt::Error);
+
+    fn chain_with<F, C>(self, chainable: F) -> Result<T, C>
+    where
+        F: FnOnce() -> C,
+        C: ChainFail,
+    {
+        self.map_err(|e| chainable().chain(e))
     }
-    errors {
-        SpawnFailed(cmd: Vec<OsString>) {
-            description("Spawn failed")
-            display(
-                "{}: (command `{}` failed to run)",
-                ERROR_PREFIX,
-                format_cmd(cmd),
-            )
-        }
-        AssertionFailed(cmd: Vec<OsString>) {
-            description("Assertion failed")
-            display(
-                "{}: (command `{}` failed)",
-                ERROR_PREFIX,
-                format_cmd(cmd),
-            )
-        }
-        StatusMismatch(expected: bool, out: String, err: String) {
-            description("Wrong status")
-            display(
-                "Expected to {}\nstatus={}\nstdout=```{}```\nstderr=```{}```",
-                expected = if *expected { "succeed" } else { "fail" },
-                got = if *expected { "failed" } else { "succeeded" },
-                out = out,
-                err = err,
-            )
+}
+
+impl<T> ResultChainExt<T> for Result<T, io::Error> {
+    fn chain<C>(self, chainable: C) -> Result<T, C>
+    where
+        C: ChainFail,
+    {
+        self.map_err(|e| chainable.chain(e))
+    }
+
+    fn chain_with<F, C>(self, chainable: F) -> Result<T, C>
+    where
+        F: FnOnce() -> C,
+        C: ChainFail,
+    {
+        self.map_err(|e| chainable().chain(e))
+    }
+}
+
+/// Failure when processing assertions.
+#[derive(Debug)]
+pub struct AssertionError {
+    cmd: Vec<ffi::OsString>,
+    cause: Option<failure::Error>,
```

Comment:
The `#[fail(cause)]` doesn't allow `cause` to be an `Option` or a `failure::Error`, so had to drop the usage of the derive macro.

Reply:
Admittedly I've only glanced at this but I'll throw it out there: have you considered the ErrorKind pattern here? One variant could declare a specific cause, another without, e.g.

```rust
enum AssertionErrorKind {
     IoError(#[cause] std::io::Error),
     StatusError { unexpected: bool, ... },
     ExitCodeError { ... }
}

```

Reply:
The encouragement within `Failure` is that `Kind` should [contain no data](https://boats.gitlab.io/failure/error-errorkind.html#what-should-your-errorkind-contain):

> Your error kind probably should not carry data - and if it does, it should only carry stateless data types that provide additional information about what the ErrorKind means. This way, your ErrorKind can be Eq, making it easy to use as a way of comparing errors.

Maybe this would be the exception it mentions

On the other hand, one of the issues is ensuring that we add all of the relevant data to an error.  A lot of this data is not available at once but gets added as we unwind the stack.  This PR explores doing this by making it easy to add arbitrary state as you unwind or having errors defined for each level of the stack.  The latter is what we ended up going with.